sdk/core/azure_core/src/policies/bearer_token_policy.rs

```diff
@@ -38,24 +40,14 @@ impl BearerTokenCredentialPolicy {
             .collect::<Vec<&str>>()
     }
 
-    fn refresh_token(&self, new_access_token: &AccessToken) {
-        let mut access_token = self.access_token.write().unwrap();
-        *access_token = Some(new_access_token.clone());
-    }
-
-    fn is_token_expired(&self) -> bool {
-        let access_token = self.access_token.read().unwrap().clone();
-        match access_token {
-            Some(access_token) => access_token.is_expired(Some(DEFAULT_REFRESH_TIME)),
-            None => true,
-        }
-    }
-
-    fn access_token(&self) -> Result<String, &'static str> {
-        let access_token = self.access_token.read().unwrap().clone();
-        match access_token {
-            Some(access_token) => Ok(String::from(access_token.token.secret())),
-            None => Err("access_token is None."),
+    async fn access_token(&self) -> Result<String> {
```

Comment:
Why is this returning a `Result` instead of an `Option`? Can the user do anything with the info? There's a lot of complication on here converting to and from different formats just to return a `String` from a `String`.

Could just be as simple as:

```rust
pub struct BearerTokenCredentialPolicy {
    access_token: RwLock<Option<String>>,
}

impl BearerTokenCredentialPolicy {
    pub fn access_token(&self) -> Option<String> {
        let access_token = self.access_token.read().ok()?;
        access_token.as_ref().map(|s| s.clone())
    }
}
```

Also not sure why this needs to be an `Arc` since it's internal. The policy itself will be in an `Arc` to be potentially shared in multiple pipelines, but that will be a shared object. Its fields are naturally shared.

Reply:
Hi Heath, thanks for the feedback! I chose to return `Result` here so that we could use the `?` operator in `send`, in case we do in fact fail to fetch the access token we can bubble up the error to the caller.

I can try to take a refactor to return `Option` here, but then in that case we would have to "unwrap" the option in `send` and handle the error case, which is all possible but might bloat up `send` when we could instead do the error handling in this separate function `access_token`?

In response to why it is `Arc`, that is because `RwLock` does not derive the `Clone` trait. 
sample output from compiler when using `RwLock<Option<AccessToken>>`:
```
15 |     access_token: RwLock<Option<AccessToken>>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `RwLock<std::option::Option<AccessToken>>`

```

Reply:
Then the underlying error should be returned, not that the `Option` is `None`. That's not an error, really. Either the error needs to be retained (not ideal), or returning an `Option` is better. The fact an internal option is `None` is not actionable and, therefore, useless to callers.

There are lots of functions to map `Result` <==> `Option`. Search for "Result" and "Option" in <https://doc.rs/std>.